File: client/protocol.js
const { HEADER_SIZE, CRC_SIZE, MSG_NAMES } = require("./config");

const DELIMITER = Buffer.from([0xaa, 0x55]);

const CRC_TABLE = (() => {
  const table = new Uint32Array(256);
  for (let i = 0; i < 256; i++) {
    let crc = i;
    for (let j = 0; j < 8; j++) {
      crc = crc & 1 ? (crc >>> 1) ^ 0xedb88320 : crc >>> 1;
    }
    table[i] = crc >>> 0;
  }
  return table;
})();

function calcCrc32(buf) {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    crc = CRC_TABLE[(crc ^ buf[i]) & 0xff] ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

function packFrame(msgType, seq, payload = "") {
  if (typeof payload === "string") {
    payload = Buffer.from(payload, "utf-8");
  }
  const length = payload.length;
  const header = Buffer.alloc(HEADER_SIZE);
  DELIMITER.copy(header, 0);
  header[2] = msgType;
  header.writeUInt16BE(seq, 3);
  header.writeUInt16BE(length, 5);

  const typeBuf = Buffer.alloc(1);
  typeBuf[0] = msgType;
  const seqBuf = Buffer.alloc(2);
  seqBuf.writeUInt16BE(seq, 0);
  const lenBuf = Buffer.alloc(2);
  lenBuf.writeUInt16BE(length, 0);
  const crcData = Buffer.concat([typeBuf, seqBuf, lenBuf, payload]);
  const crc = calcCrc32(crcData);
  const crcField = Buffer.alloc(CRC_SIZE);
  crcField.writeUInt32BE(crc, 0);

  return Buffer.concat([header, payload, crcField]);
}

function unpackFrame(data) {
  const minFrame = HEADER_SIZE + CRC_SIZE;
  if (data.length < minFrame) {
    return null;
  }
  if (data[0] !== 0xaa || data[1] !== 0x55) {
    return null;
  }
  const msgType = data[2];
  const seq = data.readUInt16BE(3);
  const length = data.readUInt16BE(5);
  const totalSize = HEADER_SIZE + length + CRC_SIZE;

  if (data.length < totalSize) {
    return null;
  }

  const payload = data.subarray(HEADER_SIZE, HEADER_SIZE + length);
  const crcReceived = data.readUInt32BE(HEADER_SIZE + length);

  const typeBuf = Buffer.alloc(1);
  typeBuf[0] = msgType;
  const seqBuf = Buffer.alloc(2);
  seqBuf.writeUInt16BE(seq, 0);
  const lenBuf = Buffer.alloc(2);
  lenBuf.writeUInt16BE(length, 0);
  const crcData = Buffer.concat([typeBuf, seqBuf, lenBuf, payload]);
  const crcComputed = calcCrc32(crcData);

  const remaining = data.subarray(totalSize);

  return { msgType, seq, payload, valid: crcReceived === crcComputed, remaining };
}

function typeName(msgType) {
  return MSG_NAMES[msgType] || `UNKNOWN(0x${msgType.toString(16).padStart(2, "0")})`;
}

module.exports = { packFrame, unpackFrame, typeName };

File: server/protocol.py
import struct
import binascii
from config import DELIMITER, HEADER_SIZE, CRC_SIZE, MSG_NAMES


def calc_crc32(data):
    return binascii.crc32(data) & 0xFFFFFFFF


def pack_frame(msg_type, seq, payload=b""):
    if isinstance(payload, str):
        payload = payload.encode("utf-8")
    length = len(payload)
    header = struct.pack(">2sBHH", DELIMITER, msg_type, seq, length)
    crc = calc_crc32(msg_type.to_bytes(1, "big") + struct.pack(">HH", seq, length) + payload)
    return header + payload + struct.pack(">I", crc)


def unpack_frame(data):
    if len(data) < HEADER_SIZE + CRC_SIZE:
        return None

    delimiter = data[0:2]
    if delimiter != DELIMITER:
        return None

    msg_type, seq, length = struct.unpack(">BHH", data[2:7])
    total_size = HEADER_SIZE + length + CRC_SIZE

    if len(data) < total_size:
        return None

    payload = data[7:7 + length]
    crc_received = struct.unpack(">I", data[7 + length:7 + length + 4])[0]
    crc_data = msg_type.to_bytes(1, "big") + struct.pack(">HH", seq, length) + payload
    crc_computed = calc_crc32(crc_data)

    remaining = data[total_size:]

    return msg_type, seq, payload, crc_received == crc_computed, remaining


def type_name(msg_type):
    return MSG_NAMES.get(msg_type, f"UNKNOWN(0x{msg_type:02X})")


def recv_exact(sock, n):
    buffer = b""
    while len(buffer) < n:
        chunk = sock.recv(n - len(buffer))
        if not chunk:
            return None
        buffer += chunk
    return buffer
